armi/runLog.py

```diff
@@ -226,265 +195,346 @@ def setVerbosity(self, levelInput):
-                warning("Failed to concatenate logs due to IOError.")
-                error(ee)
-        elif context.MPI_RANK > 0 and context.MPI_SIZE > 1:
-            for fileObj in [self._outputStream, self._errStream]:
-                fileObj.flush()
-                fileObj.close()
-        self._restoreStandardStreams()
-
-    def _emit(self, msg):
-        print(msg)
-
-    def concatenateLogs(self):
-        """
-        Concatenate the armi run logs and delete them.
+        if self._mpiRank != 0:
+            # init stderr intercepting logging
+            filePath = os.path.join(
+                "logs", RunLog.STDERR_NAME.format(name, self._mpiRank)
+            )
+            self.stderrLogger = logging.Logger("ARMI_ERROR")
+            h = logging.FileHandler(filePath)
+            fmt = "%(message)s"
+            form = logging.Formatter(fmt)
+            h.setFormatter(form)
+            h.setLevel(logging.WARNING)
+            self.stderrLogger.handlers = [h]
+            self.stderrLogger.setLevel(logging.WARNING)
+
+            # force the error logger onto stderr
+            self.initialErr = sys.stderr
+            sys.stderr = self.stderrLogger
 
-        Should only ever be called by master.
-        """
-        info("Concatenating {0} standard streams".format(context.MPI_SIZE))
-        for rank in range(context.MPI_SIZE):
-            if not rank:
-                # skip log 0
-                continue
-            stdoutName = os.path.join("logs", _stdoutName.format(self.name, rank))
-            stderrName = os.path.join("logs", _stderrName.format(self.name, rank))
-
-            for streamName, stream, logName in zip(
-                ["STDOUT", "STDERR"],
-                [self._outputStream, self._errStream],
-                [stdoutName, stderrName],
-            ):
-                with open(logName) as logFile:
-                    data = logFile.read()
-                if data:
-                    # only write if there's something to write.
-                    rankId = "\n{0} RANK {1:03d} {2} {3}\n".format(
-                        "-" * 10, rank, streamName, "-" * 60
-                    )
-                    stream.write(rankId)
-                    stream.write(data)
-                try:
-                    os.remove(logName)
-                except OSError:
-                    warning("Could not delete {0}".format(logName))
 
+def close():
+    """End use of the log. Concatenate if needed and restore defaults"""
+    if context.MPI_RANK == 0:
+        try:
+            concatenateLogs()
+        except IOError as ee:
+            warning("Failed to concatenate logs due to IOError.")
+            error(ee)
+    else:
+        if LOG.stderrLogger:
+            _ = [h.close() for h in LOG.stderrLogger.logger.handlers]
+            LOG.stderrLogger = None
+        if LOG.logger:
+            _ = [h.close() for h in LOG.logger.logger.handlers]
+            LOG.logger = None
```

Comment:
Do we not need to make sure that all of the >0 ranks have closed their handlers before rank 0 attempts to concatenate? This may require proper MPI comms...

Reply:
Okay, I'll need your opinion on this one.

All of these `close()` statements are called inside `operatorMPI.py`: that's where all the MPI comms are being done right now.  (As far as I can tell.)

The logic inside `OperatorMPI.operate()` is clearly setting up the run so the parent process finishes at earliest 1 second after all the child processes are finished.

So... in my mind, this method is safe. Thoughts?

The problem I have is I don't want to move the cross-process communication into the logger code, that doesn't seem like the right place for it. But you haven't given me a lot of confidence in `operatorMPI.py`.